### PR TITLE
Fix suppressed exceptions not showing up in test reports

### LINKED ISSUE #14245: Suppressed exceptions of a custom exception thrown by test are not captured in the test report and build scan
This is a followup to https://github.com/gradle/gradle/issues/9487 . That issue was fixed but specifically for a use of `org.junit.jupiter.api.Assertions.assertAll()`. 
The problem is the issue still persist for custom exceptions with suppressed ones thrown from the test.
That is, the test report does not contain suppressed exceptions if the main exception is a custom one.

### Reproducer
Given following example
```
import org.junit.jupiter.api.*;

public class SuppressedExceptionsAccidentallyThrownNotShownByGradleTest {

    public static class CustomException extends RuntimeException {
        public CustomException(String message) {
            super(message);
        }
    }

    private static String toUpper(String input) {
        if (input.contains("aa")) {
            RuntimeException exception = new RuntimeException("This is an exception with suppressed one.");
            exception.addSuppressed(new RuntimeException("I am suppressed"));
            throw exception;
        }
        if (input.contains("bb")) {
            RuntimeException exception =
                    new CustomException("This is a CUSTOM exception with suppressed INVISIBLE one!");
            exception.addSuppressed(new RuntimeException("I AM SUPPRESSED AND INVISIBLE"));
            throw exception;
        }

        return input.toUpperCase();
    }

    @Test
    void okTest() {
        Assertions.assertEquals("OK", toUpper("ok"));
    }

    @Test
    void failingWithSuppressedExceptionTest() {
        Assertions.assertEquals("AA", toUpper("aa"));
    }

    @Test
    void failingWithInvisibleSuppressedExceptionTest() {
        Assertions.assertEquals("BB", toUpper("bb"));
    }
}
```

### Expected Behavior
The report for `failingWithInvisibleSuppressedExceptionTest` should contain the suppressed exception

### Current Behavior

The report for `failingWithInvisibleSuppressedExceptionTest` does not contain the suppressed exception (while the report for `failingWithSuppressedExceptionTest` does) as one can see from a screenshot:
image

### Context

Consider a suite of tests and one test is intermittently failing with an exception that has a suppressed one. The motivation is that the suppressed exception can contain important information essential for understanding why the test failed. Additionally, the information in the stacktrace of the main exception can be misleading without the suppressed one.

Consider a use in reactive streams where the main exception stacktrace may not contain lines of your code at all, but for troubleshooting purposes frameworks add the essential information as suppressed exceptions. 

For instance, under certain scenario, when an unexpected exception is thrown in Spring Project Reactor's Flux, the resulting exception's stacktrace would contain no information about your code and from where you were calling it from your test. But Spring adds the information as a suppressed exception at https://github.com/reactor/reactor-core/blob/master/reactor-core/src/main/java/reactor/core/publisher/BlockingSingleSubscriber.java#L139 and the resulting stracktrace ends with 
```
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:834)
	Suppressed: java.lang.Exception: #block terminated with an error
		at reactor.core.publisher.BlockingSingleSubscriber.blockingGet(BlockingSingleSubscriber.java:139)
		at reactor.core.publisher.Mono.block(Mono.java:1704)
		at com.appdynamics.q.q.MyTest$test$$inlined$shouldThrow$lambda$1.invoke(MyTest.java:56)
		at com.appdynamics.q.s.MyTest$test$$inlined$shouldThrow$lambda$1.invoke(MyTest.java:30)
		at com.intellij.rt.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:33)
                at com.intellij.rt.junit.JUnitStarter.main(JUnitStarter.java:58)
```
and without the suppressed stracktrace you wouldn't even know `MyTest` failed at lines `30` and `56`.

### Your Environment
Gradle 6.6
Java 11

### PR DESCRIPTION
**Pull request by melix**
This commit fixes the fact that custom exception types may also
have suppressed exceptions. Suppressed exceptions were handled
for known types but not custom exception types (which are
converted to placeholder exceptions).

Fixes #14245

### CODE REVIEW

**Comment by melix** [comment]
@bot-gradle test this

**Comment by melix** [comment]
@bot-gradle test this

File: subprojects/messaging/src/main/java/org/gradle/internal/serialize/ExceptionPlaceholder.java [hunk 1]
```diff
@@ -207,6 +207,7 @@ public Object transform(Object obj) {
         }
 
         try {
+            System.out.println("Trying to reconstruct = " + type);
```
Comment (ljacomet) [inline subprojects/messaging/src/main/java/org/gradle/internal/serialize/ExceptionPlaceholder.java:210]: This should be removed

File: subprojects/messaging/src/main/java/org/gradle/internal/serialize/ExceptionPlaceholder.java [hunk 2]
```diff
@@ -248,10 +244,22 @@ public Object transform(Object obj) {
             placeholder = new DefaultMultiCauseException(message, causes);
         }
         placeholder.setStackTrace(stackTrace);
+        System.out.println("Hello");
```
Comment (ljacomet) [inline subprojects/messaging/src/main/java/org/gradle/internal/serialize/ExceptionPlaceholder.java:247]: This as well

Comment (melix) [reply subprojects/messaging/src/main/java/org/gradle/internal/serialize/ExceptionPlaceholder.java]: woot, thanks :100:

**Pull request merged**